app/repositories/contact_match_repository.js

```diff
@@ -0,0 +1,212 @@
+        el["matched_user_id"] = userId;
+      }
+      return el;
+    });
+  }
+
+  async matchUserWithContactOnUpdate(userId, phone, email, oldPhone, oldEmail) {
+    if (!((email && oldEmail) || (phone && oldPhone))) {
+      return;
+    }
+    const records = [];
+    const query = { $or: [] };
+    phone &&
+      query.$or.push({ [`phone.value`]: phone }, { [`phone.value`]: oldPhone });
+    email &&
+      query.$or.push({ [`email.value`]: email }, { [`email.value`]: oldEmail });
+
+    await this.#getRecords(records, query);
+    if (!records.length) {
+      return;
+    }
+
+    for (let i = 0; i < records.length; i++) {
+      const r = records[i];
+      const updateParam = {};
+
+      if (!(r.email && r.phone)) {
+        continue;
+      }
+
+      email &&
+        (updateParam["email"] = await this.#updateFieldOnUpdate(
+          r.email,
+          userId,
+          email,
+          oldEmail
+        ));
+      phone &&
+        (updateParam["phone"] = await this.#updateFieldOnUpdate(
+          r.phone,
+          userId,
+          phone,
+          oldPhone
+        ));
+
+      await Contact.updateOne({ _id: r._id.toString() }, { $set: updateParam });
+    }
+  }
+
+  async #updateFieldOnDelete(data, value) {
+    return data.map((el) => {
+      if (el.value === value) {
+        delete el["matched_user_id"];
+      }
+      return el;
+    });
+  }
+
+  async matchUserWithContactOnDelete(userId, phone, email) {
+    if (!(email && phone)) {
+      return;
+    }
+    const records = [];
+    const query = { $or: [] };
+    phone && query.$or.push({ [`phone.value`]: phone });
+    email && query.$or.push({ [`email.value`]: email });
+
+    await this.#getRecords(records, query);
+    if (!records.length) {
+      return;
+    }
+
+    for (let i = 0; i < records.length; i++) {
+      const r = records[i];
+      const updateParam = {};
+
+      if (!(r.email && r.phone)) {
+        continue;
+      }
+
+      email &&
+        (updateParam["email"] = await this.#updateFieldOnDelete(
+          r.email,
+          email
+        ));
+      phone &&
+        (updateParam["phone"] = await this.#updateFieldOnDelete(
+          r.phone,
+          phone
+        ));
+
+      await Contact.updateOne({ _id: r._id.toString() }, { $set: updateParam });
+    }
+  }
+
+  async matchedContactWithUser(userData) {
+    const fields = [];
+    userData.email && fields.push("email");
+    userData.phone && fields.push("phone");
+
+    if (!fields.length) {
+      return;
+    }
+
+    const findedUsersList = await User.findAll(
+      {
+        $or: fields.map((field) => {
+          return { [field]: { $in: userData[field].map((el) => el.value) } };
+        }),
+      },
+      ["_id", ...fields]
+    );
+
+    for (const field of fields) {
+      const findedUsersObj = {};
+      for (let obj of findedUsersList) {
+        findedUsersObj[obj[field]] = obj._id;
+      }
+
+      userData[field] = userData[field].map((obj) => {
```

Comment:
it will work w/o re-assignment, e.g.

```
     userData[field].forEach(obj => {
         if (findedUsersObj[obj.value]) {
           obj["matched_user_id"] = findedUsersObj[obj.value];
         }
       });
```